src/lib.rs

```diff
@@ -0,0 +1,291 @@
+            scan_env_vars(container.env, &secret_scanner, container.name)
+        {
+            findings.extend(secrets_in_container);
+        }
+    }
+
+    if let Some(init_containers) = pod_spec.init_containers {
+        for container in init_containers {
+            if let Some(secrets_in_container) =
+                scan_env_vars(container.env, &secret_scanner, container.name)
+            {
+                findings.extend(secrets_in_container);
+            }
+        }
+    }
+
+    if let Some(ephemeral_containers) = pod_spec.ephemeral_containers {
+        for container in ephemeral_containers {
+            if let Some(secrets_in_container) =
+                scan_env_vars(container.env, &secret_scanner, container.name)
+            {
+                findings.extend(secrets_in_container);
+            }
+        }
+    }
+
+    if !findings.is_empty() {
+        return kubewarden::reject_request(
+            Some(format!(
+                "The following secrets were found in environment variables -> {}",
+                create_error_message(findings)
+            )),
+            None,
+            None,
+            None,
+        );
+    }
+
+    kubewarden::accept_request()
+}
+
+fn scan_env_vars(
+    env_vars: Option<Vec<EnvVar>>,
+    secret_scanner: &SecretScanner,
+    container_name: String,
+) -> Option<HashSet<SecretRejected>> {
+    let mut findings: HashSet<SecretRejected> = HashSet::new();
+
+    if let Some(env_vars) = env_vars {
+        for env_var in env_vars {
+            if let Some(value) = env_var.value {
+                findings.extend(scan_env_var(
+                    value.as_bytes().to_vec(),
+                    secret_scanner,
+                    env_var.name,
+                    container_name.clone(),
+                ));
+            }
+        }
+    }
+
+    if findings.is_empty() {
+        None
+    } else {
+        Some(findings)
+    }
+}
+
+fn scan_env_var(
+    input: Vec<u8>,
+    ss: &SecretScanner,
+    key: String,
+    container: String,
+) -> HashSet<SecretRejected> {
+    let mut findings: HashSet<SecretRejected> = HashSet::new();
+    let lines = input.split(|&x| (x as char) == '\n');
+
+    for (_, new_line) in lines.enumerate() {
+        let results = ss.matches(new_line);
+        for (r, matches) in results {
+            let mut strings_found: Vec<String> = Vec::new();
+            for m in matches {
+                let result = ASCII
+                    .decode(&new_line[m.start()..m.end()], DecoderTrap::Ignore)
+                    .unwrap_or_else(|_| "<STRING DECODE ERROR>".parse().unwrap());
+                strings_found.push(result);
+            }
+            if !strings_found.is_empty() {
+                findings.insert(SecretRejected {
+                    reason: r.to_string(),
+                    key: key.clone(),
+                    container: container.clone(),
+                });
+            }
+        }
+    }
```

Comment:
I would rewrite the code block like that:

```rust
for line in lines {
  let mut reasons_counter: HashMap<String, u32> = HashMap::new();
  let matches_map = ss.matches(&line);
  for (reason, match_iterator) in matches_map {
    for matchojb in match_iterator {
       if reasons_counter.contains_key(reason) {
          reasons_counter[reason] += 1
       } else {
          reasons_counter[reason] = 0
       }
    }
    if !reasons_counter.is_empty() {
      let reasons: Vec<String> = reasons_counter
         .iter()
         .map(|reason, counter| format!("{} ({} found)", reason, counter))
         .collect();
      findings.insert(SecretRejected {
        reason: reasons.join(", "),
        key: key.clone(),
        container: container.clone(),
      });
    }
  }
}
```

I don't know if that will compile, this is just to give you an idea

Reply:
I find very unlikely that an env var is rejected for two or more different reasons. Even if it does I would prefer to have two different `SecretRejected` as it would be clearer to the user. Second reason might be missed by the user if we just use a `,` (keep in mind a reason might already include a comma [example](https://github.com/newrelic/rusty-hog/blob/master/src/default_rules.json#L65))
This is how rusty-hog is doing it. I took this code from [rusty-hog](https://github.com/newrelic/rusty-hog/blob/master/src/bin/duroc_hog.rs#L312-L336)